MH-12651: Recurring scheduling efficiency fixes

gregorydlogan opened the pull request:
Per the description in MH-12651, this ticket moves a bunch of the recurring scheduling logic out of the index service(!) and into the scheduler where it belongs.  This also allows us to be much more efficient with our conflict checks, generating a single large check rather than a separate smaller one for each event.

*Work sponsored by UCT*

gregorydlogan committed:
MH-12651: Change to recurring events to shorten scheduling times.  Rather than treating each child event as a separate event, a single conflict check is made.  This conflict check is still expensive, but much less so than doing individual checks.

gregorydlogan commented on modules/matterhorn-index-service/src/test/java/org/opencastproject/index/service/impl/IndexServiceImplTest.java:
```diff
@@ -975,11 +1029,58 @@ public void updatePresenters() throws IOException, org.osgi.service.cm.Configura
 
+  //NOTE: Do not modify this without making the same modifications to the copy of this method in SchedulerServiceImpl
+  //I would have moved this to an abstract class in the scheduler-api bundle, but that would introduce a circular dependency :(
+  public List<Period> calculatePeriods(RRule rrule, Date start, Date end, long duration, TimeZone tz) {
```
I want so badly to move this to an AbstractSchedulerService where it belongs, but that class can't live in scheduler-api without introducing a circular dependency with message-broker-api.  Improvements or suggestions welcome.

JamesUoM commented on modules/matterhorn-scheduler-remote/src/main/java/org/opencastproject/scheduler/remote/SchedulerServiceRemoteImpl.java:
```diff
@@ -783,6 +855,55 @@ public ReviewStatus getReviewStatus(String mediapackageId)
+    final TimeZone timeZone = TimeZone.getDefault();
+    final TimeZone utc = TimeZone.getTimeZone("UTC");
+    TimeZone.setDefault(tz);
+    net.fortuna.ical4j.model.DateTime periodStart = new net.fortuna.ical4j.model.DateTime(start);
```
maybe import net.fortuna.ical4j.model.DateTime

gregorydlogan committed:
MH-12651: Incorporating review suggestions

gregorydlogan commented:
Done!

smarquard commented:
We have run into some odd issues with incorrect times showing in the Admin UI and some spurious conflict check results (since including this PR in our 4.x build). Not sure if it's related to this PR yet.

smarquard commented on modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java:
```diff
@@ -1391,6 +1548,55 @@ public int compare(ARecord o1, ARecord o2) {
+      events.add(p);
+      TimeZone.setDefault(utc);
+    }
+    TimeZone.setDefault(timeZone);
```
See MH-12177 for how changing the TimeZone is not thread-safe and can lead to problems. At minimum, this should be:

    TimeZone.setDefault(null)

to avoid the timezone being set incorrectly if 

    final TimeZone timeZone = TimeZone.getDefault();

happens to get called at a time when TimeZone has been set to UTC. This was previously fixed in 2.x / 3.x in MH-12095

smarquard replied:
https://bitbucket.org/cilt/opencast/commits/ea3cfb409a943ef6eb244c6a4bc20c7a2855ac3b?at=r/4.x

smarquard commented:
Is this calcualtePeriods in Util.java used? If not, it should be removed.

matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/Util.java:  public static List<Period> calculatePeriods(Date start, Date end, long duration, RRule rRule, TimeZone tz) {

smarquard commented:
@gregorydlogan please do not forget to include the fix for TimeZone.setDefault from this commit https://bitbucket.org/cilt/opencast/commits/ea3cfb409a943ef6eb244c6a4bc20c7a2855ac3b?at=r/4.x

gregorydlogan committed:
MH-12651: Incorporating review suggestions

gregorydlogan commented:
Comments should be resolved!  Reorganized some things, was not aware of the code in Util.java, so I moved it to the api, and removed two of the four copies of the method.
